PR: https://github.com/odoo/odoo/pull/180212

From: 9e769e1b11f22890e5245859053bc8dd31e42634
From: Claire Bretton (clbr)
Date: 2024-09-23 17:45:47

Total Changes: 180

[REF] account,*: Print & Send - partner default sending settings

Refactoring of the Print & Send:

Why ?
Invoicing communication channels have multiplied with the advent of electronic invoicing.
Clients want to receive PDF and electronic formats (UBL, Factur-X, Factura-e,...) in different ways
(email, peppol, post, ...), depending on the localization and their needs.
Mandatory B2G declarations of invoices are also multiplying.

The current Print & Send allows for that flexibility, but at the cost of becoming:

* quite messy: you have many checboxes that you can chose at the same time with different
meaning (B2B, B2G, sending method,...)
* hard to use in batch: as you can't setup invoice channel by partner, you need to chose
each time you click on Print & Send. Then some magical things happen to apply the right checkbox
to the right invoices.
* technically complicated: because some options are about format, some about how to send, about
one or more invoices.
* the current solution invites the user to "multiply the channels", by clicking everything
and send his invoices to his customers via all the channels possible. This creates a situation
where E-Invoicing creates a mess for accountants that receive the same invoice via multiple sources,
further causing potential duplicates, double payments, etc.

How ?

* Choose ONE channel of communication per partner to send his invoice (by Post, manual (download),
by Email, by Peppol)
* In the single-mode wizard it is still possible to send through multiple
channels but it is no longuer encouraged since we only pre-check the
preferred method on the partner (or email by default).
* Synchronize mandatory B2G declaration with invoice sending
* Choose if a B2B E-invoice format needs to be generated (UBL, Factur-X, Factura-e,...)
* Make invoice batching sending easy by applying selected defaults, this also mean B2G declarations
will be applied by default if relevant.
* We also added at partner level the possibility the select an invoice report template, as well as
mail templates.

Query count increase in test_mail.test_performance:
The composer creates missing partners, with the addition of the
company_dependent fields (invoice_edi_format, invoice_sending_method), it
results in more queries.

task-4070132

closes odoo/odoo#180212

Related: odoo/enterprise#70031
Related: odoo/upgrade#6505
Signed-off-by: Claire Bretton (clbr) <clbr@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account_peppol/models/__init__.py
+++ b/addons/account_peppol/models/__init__.py
@@ -3,6 +3,7 @@
 from . import account_edi_proxy_user
 from . import account_journal
 from . import account_move
+from . import account_move_send
 from . import res_company
 from . import res_config_settings
 from . import res_partner

--- a/addons/account_peppol/models/account_move.py
+++ b/addons/account_peppol/models/account_move.py
@@ -28,7 +28,7 @@ class AccountMove(models.Model):
         if any(move.peppol_move_state in {'processing', 'done'} for move in self):
             raise UserError(_("Cannot cancel an entry that has already been sent to PEPPOL"))
         self.peppol_move_state = False
-        self.send_and_print_values = False
+        self.sending_data = False
 
     @api.depends('state')
     def _compute_peppol_move_state(self):

--- a/None
+++ b/addons/account_peppol/models/account_move_send.py
@@ -0,0 +1,154 @@
+from base64 import b64encode
+from odoo import models, _
+from odoo.addons.account.models.company import PEPPOL_LIST
+from odoo.addons.account_edi_proxy_client.models.account_edi_proxy_user import AccountEdiProxyError
+
+
+class AccountMoveSend(models.AbstractModel):
+    _inherit = 'account.move.send'
+
+    # -------------------------------------------------------------------------
+    # ALERTS
+    # -------------------------------------------------------------------------
+
+    def _get_alerts(self, moves, moves_data):
+        # EXTENDS 'account'
+        alerts = super()._get_alerts(moves, moves_data)
+        if peppol_moves := moves.filtered(lambda m: 'peppol' in moves_data[m]['sending_methods']):
+            invalid_partners = peppol_moves.partner_id.commercial_partner_id.filtered(
+                lambda partner: partner.peppol_verification_state != 'valid')
+            ubl_warning_already_displayed = 'account_edi_ubl_cii_configure_partner' in alerts
+            if invalid_partners and not ubl_warning_already_displayed:
+                alerts['account_peppol_warning_partner'] = {
+                    'message': _("The following partners are not correctly configured to receive Peppol documents. "
+                                 "Please check and verify their Peppol endpoint and the Electronic Invoicing format"),
+                    'action_text': _("View Partner(s)"),
+                    'action': invalid_partners._get_records_action(name=_("Check Partner(s)")),
+                }
+            edi_modes = [move.company_id.account_edi_proxy_client_ids.filtered(lambda usr: usr.proxy_type == 'peppol').edi_mode for move in peppol_moves]
+            if any(edi_mode in ('test', 'demo') for edi_mode in edi_modes):
+                alerts['account_peppol_demo_test_mode'] = {
+                    'message': _("Peppol is in testing/demo mode."),
+                    'level': 'info',
+                }
+        return alerts
+
+    # -------------------------------------------------------------------------
+    # SENDING METHODS
+    # -------------------------------------------------------------------------
+
+    def _do_peppol_pre_send(self, moves):
+        if len(moves.company_id) == 1:
+            can_send = self.env['account_edi_proxy_client.user']._get_can_send_domain()
+            if moves.company_id.account_peppol_proxy_state not in can_send:
+                registration_wizard = self.env['peppol.registration'].create({'company_id': moves.company_id.id})
+                return registration_wizard._action_open_peppol_form(reopen=False)
+
+        for move in moves:
+            if move.peppol_move_state in ('ready', False):
+                move.peppol_move_state = 'to_send'
+
+    def _is_applicable_to_company(self, method, company):
+        # EXTENDS 'account'
+        if method == 'peppol':
+            return company.country_code in PEPPOL_LIST and company.account_peppol_proxy_state != 'rejected'
+        else:
+            return super()._is_applicable_to_company(method, company)
+
+    def _is_applicable_to_move(self, method, move):
+        # EXTENDS 'account'
+        if method == 'peppol':
+            return all([
+                self._is_applicable_to_company(method, move.company_id),
+                move.partner_id.commercial_partner_id.is_peppol_edi_format,
+                move.company_id.account_peppol_proxy_state != 'rejected',
+                move._need_ubl_cii_xml(move.partner_id.commercial_partner_id.with_company(move.company_id).invoice_edi_format)
+                or move.ubl_cii_xml_id and move.peppol_move_state not in ('processing', 'done'),
+            ])
+        else:
+            return super()._is_applicable_to_move(method, move)
+
+    def _hook_if_errors(self, moves_data, allow_raising=True):
+        # EXTENDS 'account'
+        # to update `peppol_move_state` as `skipped` to show users that something went wrong
+        # because those moves that failed XML/PDF files generation are not sent via Peppol
+        moves_failed_file_generation = self.env['account.move']
+        for move, move_data in moves_data.items():
+            if 'peppol' in move_data['sending_methods'] and move_data.get('blocking_error'):
+                moves_failed_file_generation |= move
+
+        moves_failed_file_generation.peppol_move_state = 'skipped'
+
+        return super()._hook_if_errors(moves_data, allow_raising=allow_raising)
+
+    def _call_web_service_after_invoice_pdf_render(self, invoices_data):
+        # EXTENDS 'account'
+        super()._call_web_service_after_invoice_pdf_render(invoices_data)
+
+        params = {'documents': []}
+        invoices_data_peppol = {}
+        for invoice, invoice_data in invoices_data.items():
+            if 'peppol' in invoice_data['sending_methods'] and self._is_applicable_to_move('peppol', invoice):
+                if invoice_data.get('ubl_cii_xml_attachment_values'):
+                    xml_file = invoice_data['ubl_cii_xml_attachment_values']['raw']
+                    filename = invoice_data['ubl_cii_xml_attachment_values']['name']
+                elif invoice.ubl_cii_xml_id and invoice.peppol_move_state not in ('processing', 'done'):
+                    xml_file = invoice.ubl_cii_xml_id.raw
+                    filename = invoice.ubl_cii_xml_id.name
+                else:
+                    invoice.peppol_move_state = 'skipped'
+                    continue
+
+                partner = invoice.partner_id.commercial_partner_id
+                if not partner.peppol_eas or not partner.peppol_endpoint:
+                    invoice.peppol_move_state = 'error'
+                    invoice_data['error'] = _('The partner is missing Peppol EAS and/or Endpoint identifier.')
+                    continue
+
+                if partner.peppol_verification_state != 'valid':
+                    invoice.peppol_move_state = 'error'
+                    invoice_data['error'] = _('Please verify partner configuration in partner settings.')
+                    continue
+
+                receiver_identification = f"{partner.peppol_eas}:{partner.peppol_endpoint}"
+                params['documents'].append({
+                    'filename': filename,
+                    'receiver': receiver_identification,
+                    'ubl': b64encode(xml_file).decode(),
+                })
+                invoices_data_peppol[invoice] = invoice_data
+
+        if not params['documents']:
+            return
+
+        edi_user = next(iter(invoices_data)).company_id.account_edi_proxy_client_ids.filtered(
+            lambda u: u.proxy_type == 'peppol')
+
+        try:
+            response = edi_user._make_request(
+                f"{edi_user._get_server_url()}/api/peppol/1/send_document",
+                params=params,
+            )
+        except AccountEdiProxyError as e:
+            for invoice, invoice_data in invoices_data_peppol.items():
+                invoice.peppol_move_state = 'error'
+                invoice_data['error'] = e.message
+        else:
+            if response.get('error'):
+                # at the moment the only error that can happen here is ParticipantNotReady error
+                for invoice, invoice_data in invoices_data_peppol.items():
+                    invoice.peppol_move_state = 'error'
+                    invoice_data['error'] = response['error']['message']
+            else:
+                # the response only contains message uuids,
+                # so we have to rely on the order to connect peppol messages to account.move
+                invoices = self.env['account.move']
+                for message, (invoice, invoice_data) in zip(response['messages'], invoices_data_peppol.items()):
+                    invoice.peppol_message_uuid = message['message_uuid']
+                    invoice.peppol_move_state = 'processing'
+                    invoices |= invoice
+                log_message = _('The document has been sent to the Peppol Access Point for processing')
+                invoices._message_log_batch(bodies={invoice.id: log_message for invoice in invoices})
+
+        if self._can_commit():
+            self._cr.commit()

--- a/addons/account_peppol/models/res_partner.py
+++ b/addons/account_peppol/models/res_partner.py
@@ -18,6 +18,10 @@ NON_PEPPOL_FORMAT = (False, 'facturx', 'oioubl_201', 'ciusro')
 class ResPartner(models.Model):
     _inherit = 'res.partner'
 
+    invoice_sending_method = fields.Selection(
+        selection_add=[('peppol', 'by Peppol')],
+    )
+
     peppol_verification_state = fields.Selection(
         selection=[
             ('not_verified', 'Not verified yet'),
@@ -122,8 +126,8 @@ class ResPartner(models.Model):
     def _update_peppol_state_per_company(self, vals=None):
         partners = self.env['res.partner']
         if vals is None:
-            partners = self.filtered(lambda p: all([p.peppol_eas, p.peppol_endpoint, p.ubl_cii_format, p.country_code in PEPPOL_LIST]))
-        elif {'peppol_eas', 'peppol_endpoint', 'ubl_cii_format'}.intersection(vals.keys()):
+            partners = self.filtered(lambda p: all([p.peppol_eas, p.peppol_endpoint, p.is_ubl_format, p.country_code in PEPPOL_LIST]))
+        elif {'peppol_eas', 'peppol_endpoint', 'invoice_edi_format'}.intersection(vals.keys()):
             partners = self.filtered(lambda p: p.country_code in PEPPOL_LIST)
 
         all_companies = None
@@ -142,10 +146,11 @@ class ResPartner(models.Model):
     # COMPUTE METHODS
     # -------------------------------------------------------------------------
 
-    @api.depends('ubl_cii_format')
+    @api.depends_context('company')
+    @api.depends('invoice_edi_format')
     def _compute_is_peppol_edi_format(self):
         for partner in self:
-            partner.is_peppol_edi_format = partner.ubl_cii_format not in NON_PEPPOL_FORMAT
+            partner.is_peppol_edi_format = partner.invoice_edi_format not in NON_PEPPOL_FORMAT
 
     # -------------------------------------------------------------------------
     # LOW-LEVEL METHODS
@@ -181,7 +186,7 @@ class ResPartner(models.Model):
         old_value = self.peppol_verification_state
         if (
             not (self.peppol_eas and self.peppol_endpoint)
-            or self.ubl_cii_format in NON_PEPPOL_FORMAT
+            or self.with_company(self.company_id).invoice_edi_format in NON_PEPPOL_FORMAT
         ):
             self.peppol_verification_state = False
         else:
@@ -192,8 +197,12 @@ class ResPartner(models.Model):
             else:
                 is_participant_on_network = self._check_peppol_participant_exists(participant_info, edi_identification)
                 if is_participant_on_network:
-                    is_valid_format = self._check_document_type_support(participant_info, self.ubl_cii_format)
-                    self.peppol_verification_state = 'valid' if is_valid_format else 'not_valid_format'
+                    is_valid_format = self._check_document_type_support(participant_info, self.with_company(self.company_id).invoice_edi_format)
+                    if is_valid_format:
+                        self.peppol_verification_state = 'valid'
+                        self.with_company(self.company_id).invoice_sending_method = 'peppol'
+                    else:
+                        self.peppol_verification_state = 'not_valid_format'
                 else:
                     self.peppol_verification_state = 'not_valid'
 
